linalg/auxiliary.cpp

```diff
@@ -0,0 +1,512 @@
+}
+
+/*
+  This implementation follows that in hypre, see hypre_ParCSRSubspacePrec()
+  in hypre/src/parcsr_ls/ams.c and also hypre_AMSSolve() in the same file.
+
+  hypre's default cyle (cycle 1) is "01210", ie, smooth, correct in space
+  1, correct in space 2, correct in space 1, smooth. Their space 1 is G and
+  space 2 is Pi by default.
+
+  The MFEM interface in mfem::HypreAMS though picks cycle 13, or 034515430,
+  which separates the Pi-space solve into three separate (scalar) AMG solves
+  instead of a single vector solve.
+
+  We choose below the hypre default, but we have experimented with some other
+  cycles; the short version is that they often work but the differences are
+  generally not large.
+*/
+void GeneralAMS::Mult(const Vector& x, Vector& y) const
+{
+   MFEM_ASSERT(x.Size() == y.Size(), "Sizes don't match!");
+   MFEM_ASSERT(curlcurl_op.Height() == x.Size(), "Sizes don't match!");
+
+   Vector residual(x.Size());
+   residual = 0.0;
+   y = 0.0;
+
+   // smooth
+   smoother.Mult(x, y);
+
+   // g-space correction
+   FormResidual(x, y, residual);
+   Vector gspacetemp(gradient.Width());
+   gradient.MultTranspose(residual, gspacetemp);
+   Vector gspacecorrection(gradient.Width());
+   gspacecorrection = 0.0;
+   gspacesolver.Mult(gspacetemp, gspacecorrection);
+   gradient.Mult(gspacecorrection, residual);
+   y += residual;
+
+   // pi-space correction
+   FormResidual(x, y, residual);
+   Vector pispacetemp(pi.Width());
+
+   pi.MultTranspose(residual, pispacetemp);
+
+   Vector pispacecorrection(pi.Width());
+   pispacecorrection = 0.0;
+   pispacesolver.Mult(pispacetemp, pispacecorrection);
+   pi.Mult(pispacecorrection, residual);
+   y += residual;
+
+   // g-space correction
+   FormResidual(x, y, residual);
+   gradient.MultTranspose(residual, gspacetemp);
+   gspacecorrection = 0.0;
+   gspacesolver.Mult(gspacetemp, gspacecorrection);
+   gradient.Mult(gspacecorrection, residual);
+   y += residual;
+
+   // smooth
+   FormResidual(x, y, residual);
+   Vector temp(x.Size());
+   smoother.Mult(residual, temp);
+   y += temp;
+}
+
+// Pi-space constructor
+MatrixFreeAuxiliarySpace::MatrixFreeAuxiliarySpace(
+   ParMesh& mesh_lor, Coefficient* alpha_coeff,
+   Coefficient* beta_coeff, MatrixCoefficient* beta_mcoeff, Array<int>& ess_bdr,
+   Operator& curlcurl_oper, Operator& pi,
+#ifdef MFEM_USE_AMGX
+   bool useAmgX_,
+#endif
+   int cg_iterations) :
+   Solver(pi.Width()),
+   comm(mesh_lor.GetComm()),
+   matfree(NULL),
+   cg(NULL),
+#ifdef MFEM_USE_AMGX
+   useAmgX(useAmgX_),
+#endif
+   inner_aux_iterations(0)
+{
+   H1_FECollection * fec_lor = new H1_FECollection(1, mesh_lor.Dimension());
+   ParFiniteElementSpace fespace_lor_d(&mesh_lor, fec_lor, mesh_lor.Dimension(),
+                                       Ordering::byVDIM);
+
+   // build LOR AMG v-cycle
+   if (ess_bdr.Size())
+   {
+      fespace_lor_d.GetEssentialTrueDofs(ess_bdr, ess_tdof_list);
+   }
+   ParBilinearForm a_space(&fespace_lor_d);
```

Comment:
A more descriptive name for `a_space` might be `a_lor`.

Reply:
Done.